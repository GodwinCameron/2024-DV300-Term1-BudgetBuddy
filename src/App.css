body {
  background-color: #F7F7F5 !important;
  color: #1F1E31 !important;
}

.topbar {
  background-color: #1F1E31;
  color: white;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
}

.topbar button {
  width: fit-content;
  background-color: #D2F35E;
  border-color: #D2F35E;
  color: #1F1E31;
  height: fit-content;
  padding: 10px;
}

.topbar button:hover {
  background-color: #e4fdcd;
  border-color: #D2F35E;
}

.topbar button:active {
  background-color: #D2F35E !important;
  border-color: #1F1E31 !important;
}

.card {
  border-radius: 10px;
  border: 0.5px solid #f2f2f2 !important;
  padding: 0px;
  margin: 0px;
}

.card p {
  padding: 0px;
  margin: 0px;
}

h3 {
  font-size: 18px !important;
  font-weight: bold;
}

.card-row {
  display: flex !important;
  flex-direction: row !important;
  justify-content: space-between !important;
  align-items: baseline !important;

  width: 100%;
}

.row-income {
  background-color: #e4fdcd !important;
}

.row-expense {
  background-color: #FFE2D3 !important;
}


.card-row span {
  display: flex !important;
  flex-direction: row !important;
  column-gap: 10px;
  justify-content: space-between !important;
  align-items: baseline !important;
}

.avatar {
  min-width: 40px;
  min-height: 40px;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: white;
  border-radius: 20px;
}

.amount {
  font-weight: bold;
  font-size: 18px;
}

.amount-save {
  display: flex;
  align-items: center;
  gap: 7px;
}

.amount-save span {
  font-size: 12px;
  font-weight: normal;
}

.form-row {
  display: flex;
  flex-direction: row;
  gap: 20px;
  margin: 20px 0px;
  justify-content: space-between;
  align-items: center;
}

.form-row select {
  width: 70px;
}

.add-income {
  background-color: #D2F35E !important;
  border-color: #D2F35E !important;
  color: #1F1E31 !important;
  font-weight: bold !important;
}

.add-expenses {
  background-color: #FC677A !important;
  border-color: #FC677A !important;
  font-weight: bold !important;
}

.add-income:hover {
  background-color: #e4fdcd !important;
}

.add-expenses:hover {
  background-color: #FFE2D3 !important;
  color: #1F1E31 !important;
}

.scroll-row {
  min-width: 100%;
  overflow-x: scroll;
  overflow-y: hidden;
  display: flex;
  flex-direction: row;
  gap: 20px;
  margin-top: 20px;
}

.card-col {
  width: fit-content;
  min-width: 220px !important;
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 20px;
}

.row-tax {
  background-color: #FFEED2 !important;
}

.row-savings {
  background-color: #DDDDF7 !important;
}


.total-card {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
}

.icon {
  width: 60px;
  height: 60px;
  background-color: #E1E1E1;
  display: flex;
  justify-content: center;
  align-items: center;
  border-radius: 10px;
}

.total-card .labels {
  text-align: right;
}

.title-row {
  display: flex;
  flex-direction: row;
  gap: 20px;
  justify-content: space-between;
  align-items: center;
}

.title-row span {
  display: flex;
  flex-direction: row;
  justify-content: end;
  gap: 10px;
  align-items: center;
}

.title-row span select {
  width: 90px;
}

.title {
  display: flex;
  flex-direction: row;
  align-items: center;
  gap: 20px;
}

.total-block {
  background-color: #1F1E31 !important;
  color: white !important;
}

.last-total-card {
  background-color: #1F1E31 !important;
  color: white !important;
  margin-top: 40px;
}

.expense-outer {
  max-height: 70vh;
  overflow-y: scroll;
}

.income-outer {
  max-height: 150px;
  overflow-y: scroll;
}

.final-total {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}

.final-total .icon {
  margin-top: -50px;
  margin-bottom: 20px;
}