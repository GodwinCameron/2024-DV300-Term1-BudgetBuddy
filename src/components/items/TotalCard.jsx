import React from "react";

function TotalCard(props) {
  return (
    <div data-testid="total-card" className="total-card p-2">
      {/* <div className='icon' style={{backgroundColor: total.color}}> */}
      <div className="icon" style={{ backgroundColor: "#E4FDCD" }}>
        {/* <img src={total.icon} width={40} height={40} /> */}
        <img src={"income.png"} width={40} height={40} />
      </div>
      <div className="labels">
        {props && (
          <>
            <h4>
              {new Intl.NumberFormat("en-za", {
                style: "currency",
                currency: "ZAR",
              }).format(props.total)}
            </h4>
            {/* <p>{total.label}</p> */}
          </>
        )}
        <p>Income before tax</p>
      </div>
    </div>
  );
}

export default TotalCard;
