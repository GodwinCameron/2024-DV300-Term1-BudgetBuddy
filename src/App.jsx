import Col from "react-bootstrap/Col";
import Container from "react-bootstrap/Container";
import Row from "react-bootstrap/Row";
import "./App.css";
import Income from "./components/Income";
import Expenses from "./components/Expenses";
import Taxes from "./components/Taxes";
import TotalCard from "./components/items/TotalCard";
import { dummyCards, lastTotalCard } from "./utils";
import Savings from "./components/Savings";
import LastTotalCard from "./components/items/LastTotalCard";
import { Button } from "react-bootstrap";
import { useEffect, useState } from "react";
import TotalCardCopy from "./components/items/TotalCardcopy";

function App() {
  const [income, setIncome] = useState([]);
  const [total, setTotal] = useState(0);
  const [totalTaxed, setTotalTaxed] = useState(0);
  const [savedPercentage, setSavedPercentage] = useState(0);
  const [savedArray, setSavedArray] = useState([]);
  const [expenses, setExpenses] = useState([
    { name: "Netflix", cost: 200 },
    { name: "Petrol", cost: 2500 },
  ]);
  const [finalAmount, setFinalAmount] = useState(0);

  useEffect(() => {
    const total = income.reduce((acc, curr) => {
      return acc + Number(curr.salary);
    }, 0);
    setTotal(total);
  }, [income]);

  useEffect(() => {
    const wallet = income
      .map((item) => item.salary - item.tax.amount)
      .reduce((acc, item) => acc + item, 0);
    const totalExpenses = expenses
      .map((item) => item.cost)
      .reduce((acc, item) => acc + item, 0);
    const totalSavings = savedArray
      .map((item) => item.saved)
      .reduce((acc, item) => acc + item, 0);
    setFinalAmount(wallet - totalExpenses - totalSavings);
  }, [income, expenses, savedArray]);

  const handleResetValues = () => {
    setIncome([]);
    setTotal(0);
    setTotalTaxed(0);
    setSavedPercentage(0);
    setSavedArray([]);
    setExpenses([
      { name: "Netflix", cost: 200 },
      { name: "Petrol", cost: 2500 },
    ]);
  };

  return (
    <Container fluid>
      {/* ==========================================Top Bar============================================= */}
      <Row>
        <Col xs={12} className="topbar p-3 mb-4">
          <h2 className="title">
            <img src="logo512.png" width={60} height={60} />
            <span>BudgetBuddy</span>
          </h2>
          <Button onClick={handleResetValues}>
            <img src="reset.png" width={30} height={30} />
          </Button>
        </Col>
      </Row>

      <Row>
        {/* ==========================================First Column======================================= */}
        <Col xs={12} sm={12} md={5}>
          <div className="card shadow-sm p-3 mb-4 bg-white rounded">
            <Income income={income} setIncome={setIncome} />
          </div>
          <div className="card shadow-sm p-3 mb-4 bg-white rounded">
            <Taxes
              income={income}
              totalTaxed={totalTaxed}
              setTotalTaxed={setTotalTaxed}
            />
          </div>
          <div className="card shadow-sm p-3 mb-4 bg-white rounded">
            <Savings
              savedArray={savedArray}
              setSavedArray={setSavedArray}
              savedPercentage={savedPercentage}
              setSavedPercentage={setSavedPercentage}
              income={income}
            />
          </div>
        </Col>

        {/* ==========================================Second Column======================================= */}
        <Col xs={12} md={4}>
          <div className="card shadow-sm p-3 mb-4 bg-white rounded">
            <Expenses expenses={expenses} setExpenses={setExpenses} />
          </div>
        </Col>

        {/* ==========================================Third Column======================================= */}
        <Col xs={12} md={3}>
          {total > 0 && (
            <div className="card shadow-sm p-3 mb-4 rounded">
              <TotalCard total={total} />
            </div>
          )}
          {income.length > 0 && (
            <>
              <div className="card shadow-sm p-3 mb-4 rounded">
                <TotalCardCopy
                  icon={"tax.png"}
                  income={income}
                  color={"#FFEED2"}
                  label={"Income after tax"}
                />
              </div>
              <div className="card shadow-sm p-3 mb-4 rounded">
                <TotalCardCopy
                  icon={"expense.png"}
                  income={income}
                  color={"#FFE2D3"}
                  label={"Total Expenses"}
                  expenses={expenses}
                />
              </div>
              <div className="card shadow-sm p-3 mb-4 rounded">
                <TotalCardCopy
                  icon={"savings.png"}
                  income={income}
                  color={"#DDDDF7"}
                  label={"Total Savings"}
                  savedArray={savedArray}
                />
              </div>
            </>
          )}

          <div className="card shadow-sm p-3 mb-4 rounded last-total-card">
            <LastTotalCard total={lastTotalCard} amount={finalAmount} />
          </div>
        </Col>
      </Row>
    </Container>
  );
}

export default App;
