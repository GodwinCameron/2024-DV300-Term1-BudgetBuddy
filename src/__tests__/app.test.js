import { fireEvent, render, screen } from "@testing-library/react";
import App from "../App";
import TotalCard from "../components/items/TotalCard";
import TotalCardCopy from "../components/items/TotalCardcopy";

describe("Tests for App", () => {
  test("test component for default state", () => {
    // ARRANGE

    // ACT
    render(<App />);
    var mainTitle = screen.getByText("BudgetBuddy");

    //ASSERT
    expect(mainTitle).toBeInTheDocument();
  });
});

describe("Tests for totalCard component", () => {
  test("test for totalCard deault state", () => {
    render(<TotalCard />);
    var totalCardContainer = screen.getByTestId("total-card");
    expect(totalCardContainer).toBeInTheDocument();
  });
});

describe("Tests for totalCardCopy component", () => {
  
  test("test for totalCardCopy with a basic default savedArray", () => {
    var savedArray = [{ index: 0, salary: "12500", saved: 1250 }];
    var expenses = [{ name: "Netflix", cost: 200 }, { name: "Petrol", cost: 2500 }];
    var label = "Total Expenses";
    var icon = "savings.png";
    var income = [{ icon: "👱‍♂️", name: "John", salary: 12500, tax: {amount: 2250, extra: 0, percent: 0.18} }];
    render(<TotalCardCopy avedArray={savedArray} expenses={expenses} label={label} icon={icon} income={income} />);
    var totalCardCopyContainer = screen.getByTestId("total-card-copy");
    var totalExpeseAmount = screen.getByTestId("total-expense-amount");

    expect(totalCardCopyContainer).toBeInTheDocument();
    expect(totalExpeseAmount).toBeInTheDocument();
    expect(totalExpeseAmount).toBeInTheDocument(); // Use toHaveTextContent for checking text content
});
});

describe("snapshots", () => {
  test("snapshot of App", () => {
    const { asFragment } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
  });
});