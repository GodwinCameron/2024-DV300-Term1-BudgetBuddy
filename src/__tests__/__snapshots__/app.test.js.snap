// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots snapshot of App 1`] = `
<DocumentFragment>
  <div
    class="container-fluid"
  >
    <div
      class="row"
    >
      <div
        class="topbar p-3 mb-4 col-12"
      >
        <h2
          class="title"
        >
          <img
            height="60"
            src="logo512.png"
            width="60"
          />
          <span>
            BudgetBuddy
          </span>
        </h2>
        <button
          class="btn btn-primary"
          type="button"
        >
          <img
            height="30"
            src="reset.png"
            width="30"
          />
        </button>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col-md-5 col-sm-12 col-12"
      >
        <div
          class="card shadow-sm p-3 mb-4 bg-white rounded"
        >
          <div>
            <h3>
              Household Montly Salaries
            </h3>
            <form
              class="form-row"
              data-testid="form"
            >
              <select
                autocomplete="off"
                class="form-select"
                name="icon"
              >
                <option
                  disabled=""
                  selected=""
                >
                  -
                </option>
                <option
                  value="üë©"
                >
                  üë©
                </option>
                <option
                  value="üë±‚Äç‚ôÇÔ∏è"
                >
                  üë±‚Äç‚ôÇÔ∏è
                </option>
                <option
                  value="üê∂"
                >
                  üê∂
                </option>
                <option
                  value="üê±"
                >
                  üê±
                </option>
                <option
                  value="üê∞"
                >
                  üê∞
                </option>
                <option
                  value="üêª"
                >
                  üêª
                </option>
                <option
                  value="ü¶Å"
                >
                  ü¶Å
                </option>
                <option
                  value="üê∑"
                >
                  üê∑
                </option>
                <option
                  value="üêÆ"
                >
                  üêÆ
                </option>
              </select>
              <input
                autocomplete="off"
                class="form-control"
                id="name"
                name="name"
                placeholder="Member Name"
                type="text"
              />
              <input
                autocomplete="off"
                class="form-control"
                id="income"
                name="income"
                placeholder="0.00"
                step="0.01"
                type="number"
              />
              <button
                class="add-income btn btn-primary"
                type="submit"
              >
                Add
              </button>
            </form>
            <div
              class="income-outer hide-scroll"
            />
          </div>
        </div>
        <div
          class="card shadow-sm p-3 mb-4 bg-white rounded"
        >
          <div
            data-toggle="tooltip"
            title="Those bastards!"
          >
            <h3>
              Tax Bracket Calculation
            </h3>
            <div
              class="scroll-row hide-scroll"
            />
          </div>
        </div>
        <div
          class="card shadow-sm p-3 mb-4 bg-white rounded"
        >
          <div>
            <div
              class="title-row"
            >
              <h3>
                Savings Calculation
              </h3>
              <span>
                <p>
                  % you want to save
                </p>
                <select
                  autocomplete="off"
                  class="form-select"
                  id="percentage"
                  name="percentage"
                >
                  <option
                    disabled=""
                    selected=""
                  >
                    -
                  </option>
                  <option
                    value="5"
                  >
                    5%
                  </option>
                  <option
                    value="10"
                  >
                    10%
                  </option>
                  <option
                    value="15"
                  >
                    15%
                  </option>
                  <option
                    value="20"
                  >
                    20%
                  </option>
                </select>
              </span>
            </div>
            <div
              class="scroll-row hide-scroll"
            />
          </div>
        </div>
      </div>
      <div
        class="col-md-4 col-12"
      >
        <div
          class="card shadow-sm p-3 mb-4 bg-white rounded"
        >
          <div>
            <h3>
              Household Montly Expenses
            </h3>
            <div
              class="form-row"
            >
              <input
                autocomplete="off"
                class="form-control"
                id="title"
                name="title"
                placeholder="Expense Title"
                type="text"
              />
              <input
                autocomplete="off"
                class="form-control"
                id="expense"
                name="expense"
                placeholder="0.00"
                step="0.01"
                type="number"
              />
              <button
                class="add-expenses btn btn-primary"
                type="button"
              >
                Add
              </button>
            </div>
            <div
              class="expense-outer hide-scroll"
            >
              <div
                class="card card-row p-3 mb-2 row-expense"
                data-testid="expense-row"
              >
                <p>
                  Netflix
                </p>
                <p
                  class="amount"
                >
                  R¬†200.00
                </p>
              </div>
              <div
                class="card card-row p-3 mb-2 row-expense"
                data-testid="expense-row"
              >
                <p>
                  Petrol
                </p>
                <p
                  class="amount"
                >
                  R¬†2,500.00
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-md-3 col-12"
      >
        <div
          class="card shadow-sm p-3 mb-4 rounded last-total-card"
        >
          <div
            class="final-total p-2"
          >
            <div
              class="icon"
              style="background-color: rgb(252, 58, 33);"
            >
              <img
                height="40"
                src="wallet.png"
                width="40"
              />
            </div>
            <h2
              style="color: rgb(255, 116, 94);"
            >
              -R¬†2,700.00
            </h2>
            <p>
              Total Left To Spend Each Month üò≠
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
